src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -485,22 +484,68 @@ TypeSymbol BindRecursivePatternType(TypeSyntax typeSyntax, TypeSymbol inputType,
+                if (IsZeroElementTupleType(declType))
+                {
+                    // Work around https://github.com/dotnet/roslyn/issues/20648: The compiler's internal APIs such as `declType.IsTupleType`
+                    // do not correctly treat the non-generic struct `System.ValueTuple` as a tuple type.  We explicitly perform the tests
+                    // required to identify it.  When that bug is fixed we should be able to remove this if statement.
+                    BindValueTupleSubpatterns(deconstructClause, declType, ImmutableArray<TypeSymbolWithAnnotations>.Empty, inputValEscape, ref hasErrors, patterns, diagnostics);
+                }
+                else if (declType.IsTupleType)
+                {
+                    // It is a tuple type. Work according to its elements
+                    BindValueTupleSubpatterns(deconstructClause, declType, declType.TupleElementTypes, inputValEscape, ref hasErrors, patterns, diagnostics);
+                }
+                else
+                {
+                    // It is not a tuple type. Seek an appropriate Deconstruct method.
+                    var inputPlaceholder = new BoundImplicitReceiver(deconstructClause, declType); // A fake receiver expression to permit us to reuse binding logic
+                    var deconstructDiagnostics = DiagnosticBag.GetInstance();
+                    BoundExpression deconstruct = MakeDeconstructInvocationExpression(
+                        deconstructClause.Subpatterns.Count, inputPlaceholder, deconstructClause, deconstructDiagnostics, outPlaceholders: out ImmutableArray<BoundDeconstructValuePlaceholder> outPlaceholders,
+                        out bool anyDeconstructCandidates);
+                    if (!anyDeconstructCandidates && ShouldUseITupleForRecursivePattern(node, declType, diagnostics, out var iTupleType, out var iTupleGetLength, out var iTupleGetItem))
+                    {
+                        // There was no Deconstruct, but the constraints for the use of ITuple are satisfied.
+                        // Use that and forget any errors from trying to bind Deconstruct.
+                        deconstructDiagnostics.Free();
+                        BindITupleSubpatterns(deconstructClause, patterns, diagnostics);
+                        deconstructionSubpatterns = patterns.ToImmutableAndFree();
+                        return new BoundITuplePattern(node, iTupleGetLength, iTupleGetItem, deconstructionSubpatterns, inputType, hasErrors);
+                    }
+                    else
+                    {
+                        diagnostics.AddRangeAndFree(deconstructDiagnostics);
+                    }
+
+                    deconstructMethod = BindDeconstructSubpatterns(deconstructClause, inputValEscape, deconstruct, outPlaceholders, patterns, ref hasErrors, diagnostics);
+                }
+
+                deconstructionSubpatterns = patterns.ToImmutableAndFree();
```

Comment:
>deconstructionSubpatterns = patterns.ToImmutableAndFree(); [](start = 16, length = 58)

The flow of this method would be easier to follow if `BindValueTupleSubpatterns` and `BindDeconstructSubpatterns` returned patterns (which we would store into `deconstructionSubpatterns`).
`BindDeconstructSubpatterns` would set `deconstructMethod` via a secondary output (an `out` parameter).

Reply:
They produce multiple subpatterns, not a single pattern, which is why they go into a builder.

---
In reply to: [234825978](https://github.com/dotnet/roslyn/pull/31056#discussion_r234825978) [](ancestors = 234825978)